tensorflow/compiler/xla/service/gpu/target_util.cc

```diff
@@ -18,85 +18,267 @@ limitations under the License.
 #include "tensorflow/compiler/xla/service/gpu/target_util.h"
 
 #include "llvm/IR/MDBuilder.h"
+#include "absl/strings/str_cat.h"
+#include "tensorflow/compiler/xla/primitive_util.h"
+#include "tensorflow/compiler/xla/service/llvm_ir/ir_builder_mixin.h"
 #include "tensorflow/compiler/xla/service/llvm_ir/llvm_util.h"
+#include "tensorflow/compiler/xla/xla_data.pb.h"
 #include "tensorflow/core/platform/logging.h"
 
 namespace xla {
+
+using absl::StrAppend;
+
 namespace gpu {
 namespace {
 // Utility functions to obtain NVPTX/AMDGPU specific information.
 
-// Wrapper structure for carrying llvm intrinsic ids for NVPTX/AMDGPU platforms.
-struct TargetIntrinsics {
-  llvm::Intrinsic::ID nvptx_intrinsic;
-  llvm::Intrinsic::ID amdgpu_intrinsic;
+//Wrapper structure for  intrinsic information for NVPTX/AMDGPU. 
+  struct TargetIntrinsicInfo {
+    TargetIntrinsicInfo() : intrinsic(llvm::Intrinsic::not_intrinsic) {}
+    TargetIntrinsicInfo(llvm::Intrinsic::ID x) : intrinsic(x) {}
+    llvm::Intrinsic::ID intrinsic;
+  };
+
+//Wrapper structure for  device functionc information for NVPTX/AMDGPU. 
+  struct TargetFunctionInfo {
+    TargetFunctionInfo() : callee_name(""), 
+                  input_types({ PRIMITIVE_TYPE_INVALID }),
+                  result_type(PRIMITIVE_TYPE_INVALID),
+                  use_bitcast(false) {}
+    TargetFunctionInfo(const string callee_name_) : callee_name(callee_name_), 
+                  input_types({ PRIMITIVE_TYPE_INVALID }),
+                  result_type(PRIMITIVE_TYPE_INVALID),
+                  use_bitcast(false) {}
+    TargetFunctionInfo(const string callee_name_, 
+                       absl::Span<const PrimitiveType> input_types_, 
+                       const PrimitiveType output_type_,
+                       bool use_bitcast_) : 
+                  callee_name(callee_name_), 
+                  input_types(input_types_),
+                  result_type(output_type_),
+                  use_bitcast(use_bitcast_) {}
+    // Device function name. 
+    const string callee_name;
+    // Inpute types accespted by the device function.
+    absl::Span<const PrimitiveType> input_types;
+    // Result type of the device function.
+    PrimitiveType result_type;
+    // Use bitcast to generate casts if the desired signature at the call site 
+    // does not match the signature of the device function.
+    bool use_bitcast;
+  } target_function_info; 
+
+// Wrapper structure to carry either information about the intrinsic
+// or device function for NVPTX/AMDGPU.
+struct TargetInfo {
+  struct TargetIntrinsicInfo target_intrinsic_info;
```

Comment:
@sumesh13 would you consider using `absl::optional<>` in TargetInfo? clients can then query if it's an intrinsic or a function more easily.